### PR TITLE
docs: add design page/contributor guide

### PR DESCRIPTION
**Pull request by chingor13**
Adds a brain dump of the implementation design of this library and a few pointers for contributing.

This doc is aimed at maintainers of this libraries and folks who wish to contribute to this project. It is not written for developers who want to use release-please.

### CODE REVIEW

File: docs/design.md [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+### Branch name
+
+The name of the HEAD branch that `release-please` creates its pull request from contains
+important information that `release-please` needs.
+
+As such, we implement a helper [`BranchName` class][branch-name] that encapsulates that
+data.
+
+The branch name is not customizable at this time.
```
Comment (suztomo) [inline docs/design.md:178]: Earlier, there is "Usually default branch but can be configured to any".
I felt these statements are contradicting. Can you add clarification?

Comment (chingor13) [reply docs/design.md:178]: This is the name of the HEAD branch. In GitHub terms for a pull request there is the head branch (contains the code to be merged) and the base branch (the target branch). The target branch is configurable, the name of the branch release-please pushes code to is not.

File: docs/design.md [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+[`probot`][probot] bot handler).
+
+To actually open the pull request and update the code, we leverage the
+[`code-suggester`][code-suggester] library.
+
+### Reducing API calls
+
+Where possible, we would like to cache API calls so we limit our quota usage. An example of
+this is the [`RepositoryFileCache`][repository-file-cache] which is a read-through cache
+for fetching file contents/data from the GitHub API.
```
[the remark this answers is not in the document]
Comment (chingor13) [reply docs/design.md:84]: This design is solely for the underlying `release-please` library. App configuration is not necessary to discuss in this doc.

File: docs/design.md [hunk 3]
```diff
@@ -0,0 +1,284 @@
+# Release Please Implementation Design
+
+This document aims to describe the current design of `release-please` and serve as a
+primer for contributing to this library.
+
+## Lifecycle of a release
+
+1. A commit is merged/pushed to the release branch (usually the default branch, but can
```
[the remark this answers is not in the document]
Comment (chingor13) [reply docs/design.md:8]: Adding a "General concepts" section describing both of these.

File: docs/design.md [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+We heavily rely on unit testing to ensure `release-please` is behaving as expected. This is
+a very complex codebase and we try to avoid breaking changes.
+
+**Contributor note**: If you implement a new bugfix, please also add a new corresponding
+test to ensure we don't regress in the future.
+
+## Public interface
+
+Typescript/Javascript has limitations in its visibility scopes. If you choose to organize
```
[the remark this answers is not in the document]
Comment (chingor13) [reply docs/design.md:307]: Swapped the description of the public interface with the rationale.

File: docs/design.md (hunk 1 above)
Comment (chingor13) [reply docs/design.md]: Updated to clarify that the "HEAD branch" is not customizable at this time.

File: docs/design.md [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+This document aims to describe the current design of `release-please` and serve as a
+primer for contributing to this library.
+
+## Lifecycle of a release
+
+1. A commit is merged/pushed to the release branch (usually the default branch, but can
+   be configured to be any)
+2. `release-please` opens a "release pull request" which proposes version bumps in your
+   code as well as appends release notes to your `CHANGELOG.md`.
+3. A maintainer reviews/merges the "release pull request".
```
[the remark this answers is not in the document]
Comment (chingor13) [reply docs/design.md]: Added to the general concepts section

**Review by bcoe — APPROVED** [review]
This is a great addition to release-please's documentation 👍 

One idea that jumped out at me, through reading it, other docs we could eventually write would be:

- How to write a factory for release-please.
- How to write a new release strategy for release-plese.

File: docs/design.md [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+  * Build the manifest in code
+2. Find the SHA of each of the latest released versions for each component
+  * Iterate through the latest GitHub releases (via GitHub GraphQL API)
+  * Fallback: iterate through GitHub tags on the repository
+3. Iterate backwards through commits until we've seen all the release SHAs or we hit
+   a (configurable) max number of commits. Include fetching files for each of those
+   commits
+4. Split commits for each component. Only commits that touch the directory of the
+   component apply to that component.
+5. Run any strategy pre-configurators.
```
[the remark this answers is not in the document]
Comment (chingor13) [reply docs/design.md:90]: Yes, this is already a lifecycle event on the plugin

**Pull request merged**